### ISSUE TITLE
Color Issues

### ISSUE THREAD

**Issue by jeff-hykin** [georgewfraser/vscode-tree-sitter#14]
I'm not sure how the colors of theme's are being applied but lots of themes still look worse with the tree sitter, at least for C++ 😕I'd recommend upgrading the C++ TextMate grammar to pull from [VS Code's upstream source](https://github.com/jeff-hykin/cpp-textmate-grammar) but even after that there are some bugs that need to be worked out.

I think this issue will be fixed when the TreeSitter only adds colors, and the added colors correspond to the theme's colors for those things.

## **Material Theme**:
With the tree sitter:
<img width="947" alt="material-with-sitter" src="https://user-images.githubusercontent.com/17692058/58371791-8e806a80-7ee2-11e9-8ad2-8bd16d9afb3f.png">

Without the tree sitter and with the `Better C++ Syntax` extension (the upstream for the C++ syntax):
<img width="862" alt="material-no-sitter" src="https://user-images.githubusercontent.com/17692058/58371776-537e3700-7ee2-11e9-8cf0-b731c53b719a.png">

## XD Theme
Using a theme that is more fitted to C++ make the difference more extreme:

With the tree sitter:
<img width="897" alt="with-sitter" src="https://user-images.githubusercontent.com/17692058/58371870-942a8000-7ee3-11e9-93a2-6d9afad1e8f1.png">

Without tree-sitter, and with `Better C++ Syntax`
<img width="893" alt="no-sitter" src="https://user-images.githubusercontent.com/17692058/58371868-87a62780-7ee3-11e9-8cb6-c5eb7d27f68f.png">

**Comment by georgewfraser** [comment]
I'm not really a C++ expert so it's hard for me to tell what's good and bad. I think the reason the C++ grammar doesn't look great is because I just haven't spent much time on it. 

The C++ highlighting you see is a combination of two "layers":

1. A [simplified C++ grammar](https://github.com/georgewfraser/vscode-tree-sitter/blob/0e12c0bbc97e6daded5c779fff11804735283d8b/src/cpp.tmLanguage.json) that just colors keywords and literals.
2. A [javascript function, colorCpp,](https://github.com/georgewfraser/vscode-tree-sitter/blob/0e12c0bbc97e6daded5c779fff11804735283d8b/src/extension.ts#L178) that traverses the tree-sitter syntax tree and adds tricky colors using `setDecorations`.

There are significant advantages to this two-layer strategy:
- When you first open a file, the basic colors appear right away, and the tricky colors "pop" in after about a delay of a tenth of a second or so. 
- `setDecorations` has a meaningful cost, so it should be used selectively. If we used setDecorations to color everything, you would notice delays in coloring when you scroll around large files.

We just need to invest a little time fixing up the TextMate grammar and the color function. I just pushed a change https://github.com/georgewfraser/vscode-tree-sitter/commit/8f5d623aeb2f36ff7a9fa5bea99515baaed45eff that makes it possible for `colorCpp` to use arbitrary textmate scopes. Would be great if you wanted to tackle this!

**Comment by jeff-hykin** [comment]
The grammar still seems to appear right away even with the more complex TextMate grammar, so you might want to upgrade it at least temporarily.

Me and another guys have invested quite a lot of time fixing up the C++ grammar and I'd hate for people to have to do the same work twice.

I'd love to help out mapping the TextMate scopes to the tree-sitter! When I get back to working on this I'll take a look at the `colorCpp`.

**Comment by georgewfraser** [comment]
I don't think there's an issue with the TextMate grammar. In this extension, the job of the TextMate grammar is just to color the easy parts. This is what C++ looks like if you turn off the `colorCpp` function:

<img width="565" alt="Screen Shot 2019-05-26 at 11 53 33 AM" src="https://user-images.githubusercontent.com/1369240/58385924-e7b4d080-7fac-11e9-84cd-c0436c44c5b7.png">

It's the job of `colorCpp` to do anything that isn't in that screenshot. `colorCpp` has access to the whole parse tree, so it can do very accurate syntax coloring. For example, in Go, package members are colored differently than fields:

![shadow mov](https://user-images.githubusercontent.com/1369240/58385955-52fea280-7fad-11e9-8083-69490aec6dcc.gif)

You have to write the color function very carefully to ensure fast performance on large files but it's doable.

**Comment by jeff-hykin** [comment]
Here's some of the issues with the minimal TextMate Grammar, even the easy parts
The `#errors` are highlighted incorrectly
<img width="531" alt="Screen Shot 2019-05-26 at 9 51 51 PM" src="https://user-images.githubusercontent.com/17692058/58390811-24121c00-8001-11e9-985b-55ada6f4e83a.png">

Parameters are not highlighted at all
<img width="421" alt="Screen Shot 2019-05-26 at 9 55 30 PM" src="https://user-images.githubusercontent.com/17692058/58390862-4d32ac80-8001-11e9-9d1f-9c00e54eccf4.png">

Custom literals, hex floating point numbers, and number separators are not highlighted correctly
(all of them should be red)
<img width="217" alt="Screen Shot 2019-05-26 at 9 55 34 PM" src="https://user-images.githubusercontent.com/17692058/58390886-6a677b00-8001-11e9-9e7a-a024b30c5313.png">

Names with escapes in them are broken
<img width="400" alt="Screen Shot 2019-05-26 at 9 59 01 PM" src="https://user-images.githubusercontent.com/17692058/58390915-908d1b00-8001-11e9-8ce5-e78f771d3257.png">

And many many tokens have the wrong scope/coloring:
namespace, using, class, void, semicolon, and, or, xor, typeid, sizeof, alignas, new, delete, auto, etc

**Comment by jeff-hykin** [comment]
Here's a better example (the same theme in both pictures).

The only things highlighted correctly are `return` and the integer-literals (e.g. `10`)
The `int`, `auto`, `mutable` and `;` are highlighted incorrectly, and most things are not highlighted at all
<img width="1018" alt="Screen Shot 2019-05-26 at 10 17 18 PM" src="https://user-images.githubusercontent.com/17692058/58391518-39d51080-8004-11e9-84c9-20bfa24e9078.png">


Everything in this picture (using only the TextMate grammar) is highlighted correctly
 
<img width="1030" alt="Screen Shot 2019-05-26 at 10 18 07 PM" src="https://user-images.githubusercontent.com/17692058/58391528-45283c00-8004-11e9-9e48-dbc5e1b7459e.png">

**Comment by georgewfraser** [comment]
Help me understand this part:

> And many many tokens have the wrong scope/coloring:
namespace, using, class, void, semicolon, and, or, xor, typeid, sizeof, alignas, new, delete, auto, etc

Are you saying `namespace` should be colored differently than `using`? They're both keywords and when I look at other IDEs like Visual Studio they're colored the same.

**Comment by jeff-hykin** [comment]
Yes they're functionally very different. `new`, `delete`, `and`, `or`, `xor` are operators they're more similar to the `+` operator than the `using` keyword, and in particular the `new` and `delete` operators are memory operators.

`class`, `namespace`, `struct`, `union` , `enum` are storage type declarorators.

`if`, `else`, `while`, `try` etc are control flow, them (and functions) are the only thing that can make the code execute non linearly (e.g. more than once, or not at all)

`void` , `int` , `bool` , `double` , etc are storage types

`const`, `mutable`, `static` etc are type modifiers

`template` and `typedef` are type declarators

Etc.

The differences are really important and many many themes take advantage of them

One of the main reasons I prefer VS Code over IDE's is because the colors in VS Code can do a much better job at illustratrating the meaning of the code, and you can pick to highlight what is most meaningful to you and your project.


In the second picture: type specifiers are pink `mutable`, types are yellow `auto, int, Ret, Args`, parameter definitions are orange, functions are dark blue, control flow is light blue, assignment operators are bright red, number literals are soft red, member access is dim green.

**Comment by georgewfraser** [comment]
Ok, I think I understand: right now I have all the keywords in two categories, keyword.control and keyword. Instead I should use the longer, more specific scopes like keyword.other.namespace or whatever it is, so that existing c++ color themes can pick them up and color them differently.

We’ll also need to put back a bunch of punctuation scopes for brackets and things, and it looks like entity.name.type is different than the scope c++ was using before.

**Comment by jeff-hykin** [comment]
Sounds good!

There is a large challenge around the naming though, longer names are not really sufficient. There is a TextMate naming convention that is used as a standard to get cross compatibility across languages. On the C++ TextMate repo, we've spent a lot of time with people from VS Code coordinating changes to the scopes, how they should be named, and then talking to theme makers when they change. Whenever the TextMate conventions don't cover something, we search all of the other TextMate grammars in order to find the most similar naming among all languages.

It's a non trivial task, there are 350 scopes in C++ and more are going to be added throughout the summer. Many of them are generated programmatically. That's why I recommend starting with an existing C++ grammar and then only removing the parts that the Tree-Sitter replaces.

**Comment by isavcic** [comment]
Looking at my Go code and switching through a bunch of themes with the plugin enabled, it seems like some colors are hard-coded, like the custom type name and the types in a struct.

**Comment by jeff-hykin** [comment]
Yeah, that is one of the overall current issues with the extension. Most of the colors are hardcoded because VS Code doesn't have a good way of letting extensions get theme colors.

**Comment by georgewfraser** [comment]
No I'm getting the colors from VSCode. But if you switch themes, my colors don't update until you switch to a different file. I think VSCode has added an API to get notified of theme colors, I haven't had a chance to look into it yet though.

**Comment by jeff-hykin** [comment]
I mean colors like "source.js operator.keyword.control", not the "syntax.variable" colors
